foundation/blockchain/state/state_test.go

```diff
@@ -0,0 +1,332 @@
+	state2 := newMiner(t, MINER2_ECDSA)
+	txFactory := newTxFactory(t)
+
+	txOpts := txFactory()
+
+	state1.UpsertWalletTransaction(txOpts)
+
+	// Let them interact.
+	blk, err := state1.MineNewBlock(context.Background())
+	if err != nil {
+		t.Error(err)
+		t.FailNow()
+	}
+
+	err = state2.ProcessProposedBlock(blk)
+	if err != nil {
+		t.Error(err)
+		t.FailNow()
+	}
+
+	log.Info("Done")
+}
+
+// Test_MineAndSyncBlocksNotRespectingOrder - in this scenario we will create
+// 2 Miners, mine some blocks on Miner 1, and then, provide the blocks to Miner 2,
+// but some blocks will be missing and we expect to see it raising a database.ErrChainForked.
+func Test_MineAndForceRescynError(t *testing.T) {
+
+	log, err := logger.New("TEST")
+	if err != nil {
+		t.Error(err)
+		t.FailNow()
+	}
+
+	defer log.Sync()
+
+	state1 := newMiner(t, MINER1_ECDSA)
+	state2 := newMiner(t, MINER2_ECDSA)
+	txFactory := newTxFactory(t)
+
+	// Creating some blocks.
+	blocks := make([]database.Block, 0)
+	for i := 0; i < 20; i++ {
+		txOpts := txFactory()
+		err = state1.UpsertWalletTransaction(txOpts)
+		if err != nil {
+			t.Error(err)
+			t.FailNow()
+		}
+
+		blk, err := state1.MineNewBlock(context.Background())
+		if err != nil {
+			t.Error(err)
+			t.FailNow()
+		}
+		blocks = append(blocks, blk)
+	}
+
+	// Lets pass them to the 2nd miner, with exception of 2.
+	// This should trigger the ErrChainForked.
+	for i, b := range blocks {
+
+		if i == 12 {
+			err = state2.ProcessProposedBlock(b)
+			if err == nil {
+				t.Fatal("Should have failed here, fork issue")
+			} else if err == database.ErrChainForked {
+				t.Logf("Failling gracefully due to fork in blocks: %s", err.Error())
+				return
+			} else {
+				t.Fatalf("Ohoh - we gotta an unplanned error: %s", err.Error())
+			}
+		} else if i != 10 && i != 11 {
+			err = state2.ProcessProposedBlock(b)
+			if err != nil {
+				t.Fatal(err)
+			}
+		} else {
+			t.Log("This is the 10th, lets skip it and see.")
+		}
+	}
+}
+
+// Test_MineAndSyncBlocksNotRespectingOrder - in this scenario we will create
+// 2 Miners, mine some blocks on Miner 1, and then, provide the blocks to Miner 2,
+// but one blocks will be missing and we expect to see it raising an error.
+func Test_MineAndForceMissingBlock(t *testing.T) {
+
+	log, err := logger.New("TEST")
+	if err != nil {
+		t.Error(err)
+		t.FailNow()
+	}
+	defer log.Sync()
+
+	state1 := newMiner(t, MINER1_ECDSA)
+
+	state2 := newMiner(t, MINER2_ECDSA)
+
+	txFactory := newTxFactory(t)
+
+	// Creating the blocks.
+	blocks := make([]database.Block, 0)
+	for i := 0; i < 20; i++ {
+		txOpts := txFactory()
+		err = state1.UpsertWalletTransaction(txOpts)
+		if err != nil {
+			t.Error(err)
+			t.FailNow()
+		}
+		blk, err := state1.MineNewBlock(context.Background())
+		if err != nil {
+			t.Error(err)
+			t.FailNow()
+		}
+		blocks = append(blocks, blk)
+	}
+
+	//Lets pass them to the 2nd miner, with exception of 1.
+	for i, b := range blocks {
+
+		if i == 11 {
```

Comment:
Remove the use of else. Always try to avoid else

Use a naked switch instead where each case is a condition.